lib/build.ts

```diff
@@ -187,3 +194,173 @@ export async function runBuildTask(
 		});
 	});
 }
+
+//   This method attempts to calculate if we should pass the `platform` flag to Docker, or
+//   not.
+//
+///  The complexity arises from the fact that old v1 manifests do not describe architecture.
+//   In such cases, Docker will assume the image matches the platform of the current machine.
+//
+//   When Docker receives a target platform, these are the possible cases:
+//
+//      + V2 manifest/list for all images, and all images support the requested arch
+//        Note that "support" means either it is a single arch image with the correct arch
+//        or else it is a multi-arch image with one of the images matching arch
+//          -> Result: Good to go
+//
+//      + V2 manifest/list for all images, and some do not support the requested arch
+//          -> Result: Docker error
+//
+//      + Some V1 manifests, the platform flag matches what Docker guesses
+//          -> Result:  Docker will succeed if the guesses are correct or fail with
+//                      execution error if one of the guesses is wrong.
+//
+//      + Some V1 manifests, platform flag does not match what Docker guesses
+//          -> Result:  Docker will throw an error even if the images are actually the
+//                      correct architecture, because its guesses are wrong.
+//
+//   When Docker does not receive the `platform` flag, it sort of closes its eyes and hopes
+//   for the best.  It has to do this in order to maintain backwards-compatible behavior:
+//
+//      + V2 manifests for all images: If using multiarch images, Docker will
+//        infer the platform from where the build engine is running.  If using single arch
+//        images, Docker will assume the declared image arch is the correct platform.
+//          -> Result: Docker builds, with a warning if the images do not match the
+//                     current platform.
+//
+//      + Some V1 manifests:
+//          -> Result: Docker builds, assuming that the platform matches where the build
+//                     is occurring.
+//
+//      NOTE!  Even when a build succeeds, if Docker is making assumptions then it is possible that
+//      the image will cause an execution error on the target platform if the architecture is not
+//      correct.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manifest, we can simply pass the platform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close its eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
```

Comment:
We've removed the special cases around balenalib images now right? I think this comment needs updating.